lib/rules/prefer-regex-literals.js

```diff
@@ -107,53 +173,81 @@ module.exports = {
+                return result.value;
             }
-
-            return false;
+            return null;
         }
 
         /**
          * Determines whether the arguments of the given node indicate that a regex literal is unnecessarily wrapped.
          * @param {ASTNode} node Node to check.
+         * @param {Scope} scope The scope passed to getStringValue
          * @returns {boolean} True if the node already contains a regex literal argument.
          */
-        function isUnnecessarilyWrappedRegexLiteral(node) {
+        function isUnnecessarilyWrappedRegexLiteral(node, scope) {
             const args = node.arguments;
 
             if (args.length === 1 && isRegexLiteral(args[0])) {
                 return true;
             }
 
-            if (args.length === 2 && isRegexLiteral(args[0]) && isStaticString(args[1])) {
+            if (args.length === 2 && isRegexLiteral(args[0]) && getStringValue(args[1], scope)) {
                 return true;
             }
 
             return false;
         }
 
+        /* eslint-disable jsdoc/valid-types -- eslint-plugin-jsdoc's type parser doesn't support square brackets */
+        /**
+         * Returns a ecmaVersion compatible for regexpp.
+         * @param {import("../linter/linter").ParserOptions["ecmaVersion"]} ecmaVersion The ecmaVersion to convert.
+         * @returns {import("regexpp/ecma-versions").EcmaVersion} The resulting ecmaVersion compatible for regexpp.
+         */
+        function getRegexppEcmaVersion(ecmaVersion) {
+            /* eslint-enable jsdoc/valid-types -- JSDoc is over, enable jsdoc/valid-types again */
+            if (ecmaVersion === 3 || ecmaVersion === 5) {
+                return 5;
+            }
+            return ecmaVersion + 2009;
+        }
+
+        /**
+         * Ensures that String is the only variable present in all child scopes
+         * @param {Scope} scope The scope to go within and remove variables from
+         * @param {boolean} [children] Whether to iterate over children or not and if false iterate through parents
+         * @returns {Scope} The newer scope with only String present
+         */
+        function noStringScope(scope, children = true) {
+            scope.variables.filter(variable => variable.name !== "String").forEach(definedVariable => scope.set.delete(definedVariable.name));
+            if (children) {
+                for (const childScopeIndex in scope.childScopes) {
+                    if (!isNaN(+childScopeIndex)) {
+                        scope.childScopes[childScopeIndex] = noStringScope(scope.childScopes[childScopeIndex]);
+                    }
+                }
+                if (scope.childScopes.length === 0 && scope.upper) {
+                    scope.upper = noStringScope(scope.upper, false);
+                }
+            } else if (scope.upper) {
+                scope.upper = noStringScope(scope.upper, false);
+            }
+            return scope;
+        }
```

Comment:
This modifies scope objects, which are used by other rules too. Why do we need this, is it to limit the functionality of getStaticValue?

Reply:
I am pretty sure that this won’t modify other rule’s scope provided. It ensures that `ABC + 'str'` isn’t considered static when ABC is a constant.

> is it to limit the functionality of `getStaticValue`?

Yes

Reply:
It deletes variables from `set` for all scopes.  Any rule that needs `Scope#set` afterwards will be broken:

```js
/* eslint no-new-func: "error" */
/* eslint prefer-regex-literals: "error" */

new Function(""); // isn't reported

RegExp();
```

Also, produced scopes are broken: variables are deleted from `set` by they still exist in `variables`, `references` etc. This currently works for `eslint-utils` before it uses only `Scope#set`, but that's internal implementation of `eslint-utils` and we shouldn't rely on that.

Even if we manage to produce a valid Scope structure, it would be still relying on `eslint-utils` internals.

If `getStaticValue` isn't suitable for this use case, then we shouldn't use it.

Reply:
> It deletes variables from set for all scopes. Any rule that needs Scope#set afterwards will be broken

Is it possible to clone the `Scope`, then? (see https://astexplorer.net/#/gist/8488759c9ef7ab1b58ed0a4cdd368a4f/83d61e61be64a84baf3b90dd1a94cdad43dbb3d8)

Reply:
No, we shouldn't be doing that even if we manage to produce a valid scope structure. It's a hack that relies on the current implementation of a dependency we are using.